crypto/src/e2e_identity/identity.rs

```diff
@@ -42,139 +50,265 @@ impl MlsCentral {
+            .get_device_identities(client_ids)
+    }
+
+    /// From a given conversation, get the identity of the users (device holders) supplied.
+    /// Identity is only present for devices with a Certificate Credential (after turning on end-to-end identity).
+    /// If no member has a x509 certificate, it will return an empty Vec.
+    ///
+    /// Returns a Map with all the identities for a given users. Consumers are then recommended to
+    /// reduce those identities to determine the actual status of a user.
+    pub async fn get_user_identities(
+        &mut self,
+        conversation_id: &ConversationId,
+        user_ids: &[String],
+    ) -> CryptoResult<HashMap<String, Vec<WireIdentity>>> {
+        self.get_conversation(conversation_id)
+            .await?
+            .read()
+            .await
+            .get_user_identities(user_ids)
     }
 }
 
 impl MlsConversation {
-    fn get_user_identities(&self, client_ids: &[&ClientId]) -> CryptoResult<Vec<WireIdentity>> {
+    fn get_device_identities(&self, device_ids: &[ClientId]) -> CryptoResult<Vec<WireIdentity>> {
+        if device_ids.is_empty() {
+            return Err(CryptoError::ImplementationError);
+        }
         self.members()
             .into_iter()
-            .filter(|(m, _)| client_ids.contains(&&ClientId::from(&m[..])))
+            .filter(|(m, _)| device_ids.contains(&ClientId::from(&m[..])))
             .filter_map(|(_, c)| c.extract_identity().transpose())
             .collect::<CryptoResult<Vec<_>>>()
     }
+
+    fn get_user_identities(&self, user_ids: &[String]) -> CryptoResult<HashMap<String, Vec<WireIdentity>>> {
+        if user_ids.is_empty() {
+            return Err(CryptoError::ImplementationError);
+        }
+        let user_ids = user_ids.iter().map(|uid| uid.as_bytes()).collect::<Vec<_>>();
+        self.members()
+            .iter()
+            .filter_map(|(m, c)| UserId::try_from(m).ok().zip(Some(c)))
+            .filter(|(uid, _)| user_ids.contains(uid))
+            .filter_map(|(uid, c)| Some(uid).zip(c.extract_identity().transpose()))
+            .group_by(|(uid, _)| *uid)
+            .into_iter()
+            .map(|(uid, group)| {
+                let uid = String::try_from(uid);
+                let identities = group.into_iter().map(|(_, id)| id).collect::<CryptoResult<Vec<_>>>();
+                // TODO: simplify when `Result::zip` available
+                uid.and_then(|uid| identities.map(|ids| (uid, ids)))
+            })
+            .collect::<CryptoResult<HashMap<_, _>>>()
+    }
 }
 
 #[cfg(test)]
 pub mod tests {
     use crate::test_utils::*;
+    use crate::CryptoError;
     use wasm_bindgen_test::*;
 
     wasm_bindgen_test_configure!(run_in_browser);
 
     const ALICE_ANDROID: &str = "t6wRpI8BRSeviBwwiFp5MQ:a661e79735dc890f@wire.com";
     const ALICE_IOS: &str = "t6wRpI8BRSeviBwwiFp5MQ:ce3c1921aacdbcfe@wire.com";
+    const BOB_ANDROID: &str = "wjoxZL5tTzi2-8iND-HimA:2af3cbe39aed8cc5@wire.com";
```

Comment:
To prevent further warnings

```suggestion
    const ALICE_ANDROID: &'static str = "t6wRpI8BRSeviBwwiFp5MQ:a661e79735dc890f@wire.com";
    const ALICE_IOS: &'static str = "t6wRpI8BRSeviBwwiFp5MQ:ce3c1921aacdbcfe@wire.com";
    const BOB_ANDROID: &'static str = "wjoxZL5tTzi2-8iND-HimA:2af3cbe39aed8cc5@wire.com";
```